File: app/schemas/entity_relation.py
"""Pydantic schemas for EntityRelation CRUD + graph query responses."""

from pydantic import BaseModel


class EntityRelationCreate(BaseModel):
    from_entity_id: int
    to_entity_id: int
    relation_type: str
    description: str = ""


class EntityRelationUpdate(BaseModel):
    relation_type: str | None = None
    description: str | None = None


class EntityRelationResponse(BaseModel):
    id: int
    user_id: int
    from_entity_id: int
    to_entity_id: int
    relation_type: str
    description: str
    created_at: str
    updated_at: str


# --- Graph query response schemas ---


class GraphNode(BaseModel):
    """A node in the graph, corresponding to an Entity."""
    id: int
    name: str
    entity_type: str
    description: str = ""


class GraphEdge(BaseModel):
    """A directed edge in the graph (source/target for D3 compatibility)."""
    id: int
    source: int
    target: int
    relation_type: str
    description: str = ""


class GraphData(BaseModel):
    """Full graph data suitable for frontend visualization."""
    nodes: list[GraphNode]
    edges: list[GraphEdge]

File: frontend/src/types/entityRelation.ts
export interface EntityRelation {
  id: number;
  user_id: number;
  from_entity_id: number;
  to_entity_id: number;
  relation_type: string;
  description: string;
  created_at: string;
  updated_at: string;
}

export interface EntityRelationCreate {
  from_entity_id: number;
  to_entity_id: number;
  relation_type: string;
  description?: string;
}

export interface EntityRelationUpdate {
  relation_type?: string;
  description?: string;
}

// --- Graph visualization types ---

export interface GraphNode {
  id: number;
  name: string;
  entity_type: string;
  description?: string;
}

export interface GraphEdge {
  id: number;
  source: number;
  target: number;
  relation_type: string;
  description?: string;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}
